## Preprocessing

In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
appDF = pd.read_csv("Resources/charity_data.csv")
appDF.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
appDF = appDF.drop(columns=["EIN", "NAME"], axis=1)
appDF.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [4]:
# Determine the number of unique values in each column.
appDF.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [5]:
# Look at APPLICATION_TYPE value counts for binning
appDF["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

#cut off will be the T8 application type (737)
application_types_to_replace = appDF[appDF['APPLICATION_TYPE'].map(appDF['APPLICATION_TYPE'].value_counts()) < 737]
application_types_to_replace = application_types_to_replace['APPLICATION_TYPE'].unique().tolist()

# Replace in dataframe
for app in application_types_to_replace:
    appDF['APPLICATION_TYPE'] = appDF['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
appDF['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
Other     1529
T6        1216
T5        1173
T19       1065
T8         737
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
appDF['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [11]:
# You may find it helpful to look at CLASSIFICATION value counts >1
bucket = appDF[appDF['CLASSIFICATION'].map(appDF['CLASSIFICATION'].value_counts()) > 1]
bucket['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C4100        6
C1720        6
C2400        6
C1600        5
C1257        5
C1260        3
C2710        3
C0           3
C1267        2
C1256        2
C1234        2
C1246        2
C3200        2
Name: CLASSIFICATION, dtype: int64

In [12]:
# Choose a cutoff value and create a list of classifications to be replaced (C5000 - 116)
# use the variable name `classifications_to_replace`
classifications_to_replace = appDF[appDF['CLASSIFICATION'].map(appDF['CLASSIFICATION'].value_counts()) < 116]
classifications_to_replace = classifications_to_replace['CLASSIFICATION'].unique().tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    appDF['CLASSIFICATION'] = appDF['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
appDF['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other      887
C7000      777
C1700      287
C4000      194
C5000      116
Name: CLASSIFICATION, dtype: int64

In [13]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies = pd.get_dummies(appDF)
dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       1   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                     0                    0                    0   
1                     0                    1                    0   
2                     0                    0                    0   
3                     0                    1                    0   
4                     0                    1                    0   

   APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T8  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    1                    0                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 46 columns]

In [14]:
# Split our preprocessed data into our features and target arrays
temp_X = dummies.drop('IS_SUCCESSFUL', axis=1)
temp_Y = dummies['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, Y_train, Y_test = train_test_split(temp_X, temp_Y, random_state=1)

In [15]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [88]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

neuralNetwork = tf.keras.models.Sequential()

# First hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units=5, activation="relu", input_dim=45)) #input_dim means how many columns are in the data frame

# Second hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 3, activation="tanh"))

# Output layer
neuralNetwork.add(tf.keras.layers.Dense(units=1, activation = "sigmoid"))

# Check the structure of the model
neuralNetwork.summary()

Model: "sequential_23"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_74 (Dense)            (None, 5)                 230       
                                                                 
 dense_75 (Dense)            (None, 3)                 18        
                                                                 
 dense_76 (Dense)            (None, 1)                 4         
                                                                 
Total params: 252
Trainable params: 252
Non-trainable params: 0
_________________________________________________________________


In [89]:
# Compile the model
neuralNetwork.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [90]:
# Train the model
fit_model = fit_model = neuralNetwork.fit(X_train_scaled, Y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 1s 663us/step - loss: 0.6350 - accuracy: 0.6499
Epoch 2/50
804/804 [==============================] - 1s 676us/step - loss: 0.5768 - accuracy: 0.7201
Epoch 3/50
804/804 [==============================] - 1s 709us/step - loss: 0.5649 - accuracy: 0.7254
Epoch 4/50
804/804 [==============================] - 1s 691us/step - loss: 0.5606 - accuracy: 0.7265
Epoch 5/50
804/804 [==============================] - 1s 672us/step - loss: 0.5583 - accuracy: 0.7284
Epoch 6/50
804/804 [==============================] - 1s 667us/step - loss: 0.5574 - accuracy: 0.7272
Epoch 7/50
804/804 [==============================] - 1s 705us/step - loss: 0.5563 - accuracy: 0.7273
Epoch 8/50
804/804 [==============================] - 1s 666us/step - loss: 0.5555 - accuracy: 0.7283
Epoch 9/50
804/804 [==============================] - 1s 670us/step - loss: 0.5550 - accuracy: 0.7278
Epoch 10/50
804/804 [==============================] - 1s 667us/step - loss: 0.554

In [138]:
# Evaluate the model using the test data 

"""
First Attempt

# First hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units=5, activation="relu", input_dim=45))

# Second hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 4, activation="sigmoid"))

# Output layer
neuralNetwork.add(tf.keras.layers.Dense(units=3, activation = "sigmoid"))

# Check the structure of the model
neuralNetwork.summary()


"""


model_loss, model_accuracy = neuralNetwork.evaluate(X_test_scaled, Y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 3.1483 - accuracy: 0.2227 - 243ms/epoch - 905us/step
Loss: 3.1482717990875244, Accuracy: 0.2227405309677124


In [91]:
"""
Most accurate attempt

# First hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units=5, activation="relu", input_dim=45))

# Second hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 4, activation="sigmoid"))

# Output layer
neuralNetwork.add(tf.keras.layers.Dense(units=1, activation = "tanh"))

"""

model_loss, model_accuracy = neuralNetwork.evaluate(X_test_scaled, Y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5547 - accuracy: 0.7298 - 219ms/epoch - 817us/step
Loss: 0.554730236530304, Accuracy: 0.7297959327697754


In [73]:
"""
SEVERAL UNIT INPUTS aka really bad
# First hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units=50, activation="relu", input_dim=45)) #input_dim means how many columns

# Second hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 30, activation="sigmoid"))

# Output layer
neuralNetwork.add(tf.keras.layers.Dense(units=10, activation = "tanh"))
"""


model_loss, model_accuracy = neuralNetwork.evaluate(X_test_scaled, Y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 1.3125 - accuracy: 0.1298 - 236ms/epoch - 879us/step
Loss: 1.3125284910202026, Accuracy: 0.12979592382907867


# REPORT

## Recommendation / Conclusion

After trimming the data, and building a Neural-Network with several different algorithms, an accuracy of 72.9% has been achieved with the big data prediction. Higher levels of accuracy may be achieved with better configuration of the hidden layers.

I used an activation method for each different layer for the algorithm to apply several different mathematical interpretations of the data for it to potentially increase its accuracy. With the first attempts of the Neural-Network implementation, two hidden layers were used with both using Rectified Linear Unit (relu), yielding an accuracy of 22%. This is because not many different sorting methods were used with the data inside the Neural Network preventing the algorithm from making confident predictions.

After researching more optimal activation models (Sigmoid, Tanh), I implemented them as hidden layers and started to see much more promising results.

Another variable that can be changed to increase the accuracy of the model potentially further is the number of units put into each layer, further fine tuning the model to make accurate predictions. (very large amounts of units yield lower accurate results)


### Other model ideas

attempted to add a 4th layer with the output layer being a Softmax function, however that lowered the over all accuracy of the results because the data is not multi-class

##### First hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units=5, activation="relu", input_dim=45)) #input_dim means how many columns

##### Second hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 3, activation="sigmoid"))

##### Third hidden layer
neuralNetwork.add(tf.keras.layers.Dense(units = 2, activation="tanh"))

##### Output layer
neuralNetwork.add(tf.keras.layers.Dense(units=1, activation = "Softmax"))